<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jean Morrison" />

<meta name="date" content="2023-09-21" />

<title>Using Snakemake to Execute Commands</title>

<script src="site_libs/header-attrs-2.22/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Snakemake Tutorial</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Using Snakemake to Execute Commands</h1>
<h4 class="author">Jean Morrison</h4>
<h4 class="date">2023-09-21</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-09-21
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 1
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>snakemake_tutorial/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R
Markdown file created these results, you’ll want to first commit it to
the Git repo. If you’re still working on the analysis, you can ignore
this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjean997snakemaketutorialtree8f3e1ea3568bb5170be5b01c61938afb9f3f0451targetblank8f3e1eaa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/jean997/snakemake_tutorial/tree/8f3e1ea3568bb5170be5b01c61938afb9f3f0451" target="_blank">8f3e1ea</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomjean997snakemaketutorialtree8f3e1ea3568bb5170be5b01c61938afb9f3f0451targetblank8f3e1eaa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/jean997/snakemake_tutorial/tree/8f3e1ea3568bb5170be5b01c61938afb9f3f0451" target="_blank">8f3e1ea</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/basics.Rmd

Unstaged changes:
    Modified:   analysis/_site.yml
    Modified:   analysis/index.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>In this section, we will learn how to use Snakemake to execute
commands.</p>
<p>Goals for this session:</p>
<ol style="list-style-type: decimal">
<li>Use bcftools to summarize a vcf files</li>
<li>Write a Snakemake rule to run the bcftools command</li>
<li>Run Snakemake on the command line</li>
<li>Learn how to specify target files</li>
</ol>
</div>
<div id="data" class="section level2">
<h2>Data</h2>
<p>Inside the <code>data/</code> folder you will see three files.</p>
<pre><code>ls data/
chr20.vcf.gz  chr21.vcf.gz  chr22.vcf.gz</code></pre>
<p>These are vcf files which contain genotype information for 100 people
from 1000 Genomes phase 3. As you might guess from the file names, each
file corresponds to one chromosome for chromosomes 20, 21, and 22. We
can use bcftools to summarize the contents of these files. The command
below will generate some summaries and write them to a file. Run this
command in your tutorial directory.</p>
<pre><code>bcftools stats data/chr22.vcf.gz &gt; chr22_stats.txt</code></pre>
<p>If we take a look at this file, we find out that
<code>data/chr22.vcf.gz</code> contains 100 samples and about a million
variants. We also have lots of other information but this isn’t that
important for our purposes.</p>
</div>
<div id="using-snakemake-to-run-bcftools" class="section level2">
<h2>Using Snakemake to run bcftools</h2>
<div id="writing-the-rule" class="section level3">
<h3>Writing the Rule</h3>
<p>We now want to write a rule so that Snakemake could execute this
command for us. The input for the rule is the data file and the output
is the file <code>chr22_stats.txt</code>. Create a new file in your
tutorial directory called <code>Snakefile</code> and open it. I like to
use Vim so I would type <code>vim Snakefile</code>. If you are using
<code>nano</code>, you could type <code>nano Snakefile</code>. If you
are on the GreatLakes web interface, click on “+ New File” and then
click on Edit in the dropdown menu. Put the text below into your file
and then save and exit.</p>
<pre><code>rule get_stats22:
    input: &quot;data/chr22.vcf.gz&quot;
    output: &quot;results/chr22_stats.txt&quot;
    shell: &quot;mkdir -p results; bcftools stats data/chr22.vcf.gz &gt; results/chr22_stats.txt&quot;</code></pre>
<p>This rule lists the input, output, and action (<code>shell</code>)
for the process of using bcftools to summarize the vcf file. Our rule
differs slightly from the command we used in the previous section
because it puts the output into a directory called
<code>results/</code>.</p>
<p><strong>Workflow Tip:</strong> Organize the files in your workflow
into subdirectories. I like to keep the Snakefile and sometimes a few
helper files at the top level. Generally, I like to have sub-directories
called <code>code</code> where I put scripts (could be called
<code>R</code> or <code>scripts</code> etc), <code>data/</code>, and
<code>results/</code>. You can choose a different organizational scheme
that works for you but it is nice to have a system.</p>
</div>
<div id="running-snakemake" class="section level3">
<h3>Running Snakemake</h3>
<p>Activate the snakemake conda environment by typing</p>
<pre><code>conda activate snakemake</code></pre>
<p>Now we are ready to run Snakemake for the first time. Type</p>
<pre><code>snakemake -j 1 -c 1 -p </code></pre>
<p>The <code>-j 1</code> and <code>-c 1</code> flags tell Snakemake to
run one job at a time and to use one core. The <code>-p</code> flag
tells Snakemake to print out the shell commands it will use before it
runs them. We don’t need to tell it anything else because we named our
Snakefile <code>Snakefile</code> which is the default name that
Snakemake will look for. We could have named it anything else but then
we would need to use the <code>-s</code> flag to let Snakemake know what
it is called. For example, if you named it
<code>my_first_snakefile</code>, you could use the command</p>
<pre><code>snakemake -j 1 -c 1 -p -s my_first_snakefile</code></pre>
</div>
<div id="what-just-happened" class="section level3">
<h3>What Just Happened</h3>
<p>The output you get from the previous command should look like</p>
<pre><code>Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job stats:
job            count    min threads    max threads
-----------  -------  -------------  -------------
get_stats22        1              1              1
total              1              1              1

Select jobs to execute...

[Thu Sep 21 16:52:56 2023]
rule get_stats22:
    input: data/chr22.vcf.gz
    output: results/chr22_stats.txt
    jobid: 0
    reason: Missing output files: results/chr22_stats.txt
    resources: tmpdir=/tmp

mkdir -p results; bcftools stats data/chr22.vcf.gz &gt; results/chr22_stats.txt
[Thu Sep 21 16:53:00 2023]
Finished job 0.
1 of 1 steps (100%) done
Complete log: .snakemake/log/2023-09-21T165255.457024.snakemake.log</code></pre>
<p>In this output, Snakemake is telling us what is happening. First, it
identifies a target file and then uses our rules to build a DAG of jobs
to create that file. In our case, it identifies
<code>results/chr22_stats.txt</code> as the target file (more on why it
chose this coming) and then identifies that it needs to execute one rule
to create this file. It then executes the necessary rule and writes some
inforamtion to a log file.</p>
</div>
<div id="dry-run-mode--n" class="section level3">
<h3>Dry Run Mode (<code>-n</code>)</h3>
<p>A very useful command line option for Snakemake is <code>-n</code>.
This option will cause Snakemake to only do the first steps of
identifying the target, building the DAG, and making the plan of jobs to
execute without actually executing any of the jobs. I almost always run
Snakemake with <code>-n</code> before running it “for real” because this
can help me see that Snakemake is going to do what I thought it would
do. Let’s try this now.</p>
<pre><code>snakemake -n -p</code></pre>
<p>We can omit the <code>-j 1 -c 1</code> options because these only
relate to job execution. You should see</p>
<pre><code>Building DAG of jobs...
Nothing to be done (all requested files are present and up to date).</code></pre>
<p>Snakemake says there is nothing to do because the target file already
exists and the code hasn’t changed since the last time the target file
was created. We will talk more about when Snakemake will re-run a rule
in future sections. For now, delete <code>results/chr22_stats.txt</code>
and re-run the dry-run command above. You should now see different
output indicating that Snakemake will run the rule again.</p>
</div>
<div id="targets" class="section level3">
<h3>Targets</h3>
<p>In our file, we didn’t explicitly tell Snakemake what file to create,
it just guessed. You might imagine that in a more complicated workflow,
we will probably need to be explicit about what we want. Snakemake has a
hierarchy of places it will look to decide what the target is. To see
this in action, lets add a second rule to our Snakefile.</p>
<pre><code>rule get_stats21:
    input: &quot;data/chr21.vcf.gz&quot;
    output: &quot;results/chr21_stats.txt&quot;
    shell: &quot;mkdir -p results; bcftools stats data/chr21.vcf.gz &gt; results/chr21_stats.txt&quot;</code></pre>
<p><strong>Hierarchy for Determining Target File(s):</strong></p>
<ol style="list-style-type: decimal">
<li>Command line designation. Target files can be specified at the end
of the <code>snakemake</code> command.</li>
<li>Rules designated as <code>default_target</code>.</li>
<li>The first rule in the file.</li>
</ol>
<p>From this hierarchy, if we don’t make any further changes to the
Snakefile and run the command</p>
<pre><code>snakemake -n -p</code></pre>
<p>which file do you think will be the target file? Check your answer by
running a dry-run.</p>
<div id="specifying-targets-at-the-command-line" class="section level4">
<h4>Specifying targets at the command line</h4>
<p>The command</p>
<pre><code>snakemake -n -p results/chr21_stats.txt</code></pre>
<p>specifies the target files <code>results/chr21_stats.txt</code>. We
could also have multiple target files such as</p>
<pre><code>snakemake -n -p results/chr22_stats.txt results/chr21_stats.txt</code></pre>
<p>Instead of using the file name, we could also use the name of the
rule. Using</p>
<pre><code>snakemake -n -p get_stats21</code></pre>
<p>indicates that our target file is the output of the rule
<code>get_stats21</code>. This specification will take precedence over
any other way of indicating the target.</p>
<p>Try running</p>
<pre><code>snakemake -n -p results/chr19_stats.txt</code></pre>
<p>What happens?</p>
</div>
<div id="specifying-default_target" class="section level4">
<h4>Specifying <code>default_target</code></h4>
<p>Any rule can be a target rule as long as that rule does not contain
any wildcards (we will learn about these next). To indicate that a rule
is a target add the line <code>default_target: True</code> to the rule
before the <code>shell:</code> line. For example, modify the
<code>get_stats21</code> rule to read</p>
<pre><code>rule get_stats21:
    input: &quot;data/chr21.vcf.gz&quot;
    output: &quot;results/chr21_stats.txt&quot;
    default_target: True
    shell: &quot;mkdir -p results; bcftools stats data/chr21.vcf.gz &gt; results/chr21_stats.txt&quot;</code></pre>
<p>If you run</p>
<pre><code>snakemake -n -p</code></pre>
<p>You should see that <code>results/chr21_stats.txt</code> is the
target.</p>
<p>I believe that if multiple rules are specified with
<code>default_target: True</code>, Snakemake will pick the last one. You
should not do this as it will make your workflow confusing.</p>
</div>
<div id="using-the-first-rule" class="section level4">
<h4>Using the First Rule</h4>
<p>The most common way to use Snakemake is to allow it to execute the
first rule as the default. Often, this rule is treated specially and
given the name <code>all</code>. The <code>all</code> rule has an input
but no output or shell line. Add these lines <strong>to the top</strong>
of your Snakefile and remove any designations of
<code>default_target: True</code> that you added in the previous
section.</p>
<pre><code>rule all:
    input: results/chr22_stats.txt, results/chr21_stats.txt</code></pre>
<p>This will make the files listed in <code>input</code> for rule
<code>all</code> the target files.</p>
</div>
</div>
</div>
<div id="final-snakefile" class="section level2">
<h2>Final Snakefile</h2>
<p>At the end of this section, you Snakefile should look like the file
in <code>code/1.Snakefile</code></p>
<pre><code>rule all:
    input: &quot;results/chr21_stats.txt&quot;, &quot;results/chr22_stats.txt&quot;

rule get_stats22:
    input: &quot;data/chr22.vcf.gz&quot;
    output: &quot;results/chr22_stats.txt&quot;
    shell: &quot;mkdir -p results; bcftools stats data/chr22.vcf.gz &gt; results/chr22_stats.txt&quot;

rule get_stats21:
    input: &quot;data/chr21.vcf.gz&quot;
    output: &quot;results/chr21_stats.txt&quot;
    shell: &quot;mkdir -p results; bcftools stats data/chr21.vcf.gz &gt; results/chr21_stats.txt&quot;</code></pre>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
